Cassandra getCandidateCellsForSweeping

nziebart opened the pull request:
**Goals (and why)**:
Implement `CassandraKeyValueService#getCandidateCellsForSweeping` in a way that does not fetch values, yet still make efficient queries.

Fixes https://github.com/palantir/atlasdb/issues/2509

**Implementation Description (bullets)**:
- Add methods to `CqlExecutor` required to fetch timestamps
- Add classes implementing logic to fetch sweep candidates using the executor
- Refactor `getRangeOfTimestamps` to use `getCandidateCellsForSweeping`. Ultimately this method is not useful anymore, so it's now deprecated and we should work to remove it.
- Remove a bunch of classes specific to `getRangeOfTimestamps`

**Concerns (what feedback would you like?)**:
- Is `getRangeOfTimestamps` still useful? Is it right to try to get rid of it?
- Do the existing test have sufficient coverage

**Where should we start reviewing?**:
- `GetCandidateCellsForSweeping.java`

**Priority (whenever / two weeks / yesterday)**:
this week

<!---
Please remember to:
- Add any necessary release notes (including breaking changes)
- Make sure the documentation is up to date for your change
--->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/2546)
<!-- Reviewable:end -->

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutor.java:
```diff
@@ -60,66 +62,61 @@
+        long invertedTimestamp = ~startTimestampExclusive;
         CqlQuery query = new CqlQuery(
-                "SELECT column1, column2 FROM %s WHERE key = %s AND column1 = %s AND column2 > %s LIMIT %s;",
+                "SELECT column1, column2 FROM %s WHERE key = %s AND (column1, column2) > (%s, %s) LIMIT %s;",
```
You can do this? :mindblown:

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutor.java:
```diff
@@ -60,66 +62,61 @@
-                column1(previousColumn),
-                limit(limit));
-        return query.executeAndGetCells(row);
+    private CellWithTimestamp getCellForColumn1Column2(CqlRow row, byte[] key) {
```
nit: `getCellFromColumn1Column2`?

The `Column1Column2` here looks a bit nasty, but I struggle to think of better wording. Perhaps `getCellFromKeylessRow`?

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutor.java:
```diff
@@ -60,66 +62,61 @@
+        return CellWithTimestamp.of(Cell.create(rowName, columnName), timestamp);
+    }
+
+    private CellWithTimestamp getCellForKeyColumn1Column2(CqlRow row) {
```
`getCellFromRow`?

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeOfTimestampsIterator.java:
```diff
@@ -0,0 +1,89 @@
+import com.palantir.common.base.ClosableIterator;
+
+public class RangeOfTimestampsIterator implements ClosableIterator<RowResult<Set<Long>>> {
+    private final ClosableIterator<CandidateCellForSweeping> delegate;
```
`delegate` sounds like the wrong description here. `source`? Why do we maintain essentially two iterators here?

nziebart replied:
Just so we can call close on it =/

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeOfTimestampsIterator.java:
```diff
@@ -0,0 +1,89 @@
+                .shouldCheckIfLatestValueIsEmpty(false)
+                .build();
+        this.rangeRequest = rangeRequest;
+        this.delegate = keyValueService.getCandidateCellsForSweeping(tableRef, request)
```
This is pretty odd. Inside the constructor, we perform a hefty-looking KVS call. Not only that, but this is the place where we passed in `this` earlier. Is this an attempt to defer some computation?

nziebart replied:
no, we could just pass in the delegate iterator directly, but i wanted to avoid having this logic in the KVS. could move this to a static constructor?

gsheasby replied:
A static constructor sounds OK, if this class survives any redesign.

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeOfTimestampsIterator.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public boolean hasNext() {
+        return candidates.hasNext() && rangeRequest.inRange(peekNextRowName());
```
nit: Could _perhaps_ use a comment explaining why we want to exclude the case where the next row is not in the range.

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RangeOfTimestampsIterator.java:
```diff
@@ -0,0 +1,89 @@
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.common.base.ClosableIterator;
+
+public class RangeOfTimestampsIterator implements ClosableIterator<RowResult<Set<Long>>> {
```
Could definitely use a paragraph of javadoc here to explain the purpose of the class.

nziebart replied:
Yeah, this class was kind of added as an afterthought. my original plan was to remove `getRangeOfTimestamps` since it it's not really needed now that we have `getCandidateCellsForSweeping`. However it's used for some tangential reasons here an there (in tests, and in `AbstractKeyValueService#deleteRange`), so I just deprecated it instead.

nziebart replied:
was planning to do a follow up to see if we can just get rid of it entirely

nziebart replied:
On the other hand if we can't get rid of this, maybe we just want to make `GetCandidateCellsForSweeping` explicitly return full row objects, and take a start and end row. It adds more complexity there but we wouldn't need this class.

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCandidateCellsForSweepingIterator.java:
```diff
@@ -0,0 +1,69 @@
+            KeyValueService kvs,
+            CqlExecutor cqlExecutor, TableReference table,
+            CandidateCellForSweepingRequest request) {
+        this.kvs = kvs;
```
We take the whole KVS, only to pass it via a few intermediate classes, eventually into a helper class that uses one method of it. Is it feasible/cleaner to pass a lambda instead?

nziebart replied:
Yeah probably worth passing extracting an interface for just the functionality we need. Part of the problem is that the `CassandraKvsImpl` is huge, and not broken down into any components.

nziebart commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/Value.java:
```diff
@@ -56,6 +56,10 @@ public static Value createWithCopyOfData(byte[] contents, long timestamp) {
         return contents;
     }
 
+    public boolean isEmpty() {
```
not sure i follow?

gsheasby replied:
It's not something you added, but I think `Value.createWithCopyOfData(null, 1L)` will cause an NPE, whereas `Value.create(null, 1L)` doesn't.

I only spotted it because I was figuring out if `isEmpty()` could also cause an NPE (turns out it can't).

nziebart commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    private boolean fetchBatchOfRemainingCellTsPairsInLastRow() {
```
rename to `fetchBatchOfRemainingTimestampsForLastRow` for consistency

nziebart commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetEmptyLatestValues.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private Set<Cell> getSingleBatch(
+            List<CellWithTimestamps> batch) {
```
unnecessary newline

gsheasby replied:
^ still there.

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetEmptyLatestValues.java:
```diff
@@ -0,0 +1,75 @@
+    private Set<Cell> getSingleBatch(
+            List<CellWithTimestamps> batch) {
+        Map<Cell, Long> timestampsByCell = batch.stream()
+                .collect(Collectors.toMap(
```
The `.collect()` part could also be collected (pun intended) onto one line.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidateGroupingIterator.java:
```diff
@@ -103,13 +103,20 @@ private boolean isCurrentCell(CellTsPairInfo cellTs) {
 
+    private Collection<Long> toList(TLongList values) {
+        List<Long> result = Lists.newArrayListWithExpectedSize(values.size());
+        for (int i = 0; i < values.size(); i++) {
```
You can replace the `for` loop here with `values.forEach(result::add);`.

gsheasby commented on atlasdb-dbkvs/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidateGroupingIteratorTest.java:
```diff
@@ -97,13 +97,13 @@ private static CandidateCellForSweeping candidate(String rowName,
         return ImmutableCandidateCellForSweeping.builder()
                 .cell(Cell.create(bytes(rowName), bytes(colName)))
                 .isLatestValueEmpty(latestValEmpty)
                 .numCellsTsPairsExamined(numCellTsPairsExamined)
```
Compile break here

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepableCellFilter.java:
```diff
@@ -53,13 +53,13 @@ public BatchOfCellsToSweep getCellsToSweep(List<CandidateCellForSweeping> candid
                 }
             }
-            numCellTsPairsExamined = candidate.numCellsTsPairsExamined();
+            numCellTsPairsExamined += candidate.sortedTimestamps().size();
```
👏

gsheasby commented on atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractGetCandidateCellsForSweepingTest.java:
```diff
@@ -98,19 +98,52 @@ public void reportLatestEmptyValue() {
+                .put(2, 2, 10L)
+                .store();
+        boolean containsIgnoredTimestamps = getAllCandidates(conservativeRequest(PtBytes.EMPTY_BYTE_ARRAY, 100L, 100))
+                .stream().anyMatch(candidate -> candidate.sortedTimestamps().contains(-1L));
```
completely inconsequential nit: You can also use `.noneMatch`, if you want to assert that it's true.

gsheasby commented on atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractKeyValueServiceTest.java:
```diff
@@ -1103,6 +1103,47 @@ public void testGetRangeOfTimestampsReturnsAllRows() {
+                        Cell.create(row0, column0), value0_t1),
+                TEST_TIMESTAMP + 10);
+
+        RangeRequest range = RangeRequest.all().withBatchHint(1);
```
Could use a batch hint of 2 here, to guard against the weird failure case where we would have returned the later value, but had the earlier value first in the list.

gsheasby commented on atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractKeyValueServiceTest.java:
```diff
@@ -1103,6 +1103,47 @@ public void testGetRangeOfTimestampsReturnsAllRows() {
+                keyValueService.getRangeOfTimestamps(TEST_TABLE, range, TEST_TIMESTAMP + 1));
+        assertEquals(1, results.size());
+        assertArrayEquals(row0, results.get(0).getRowName());
+        assertEquals(TEST_TIMESTAMP, (long)results.get(0).getOnlyColumnValue().iterator().next());
```
checkstyle probably wants a space here: `(long) results`.

nziebart commented:
@gsheasby addressed your comments, and reworked the iterator to return rows rather than cells (which also makes it more explicit that that's what we're doing), so we can get rid of `RangeOfTimestampsIterator`

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/CandidateRowsForSweepingIterator.java:
```diff
@@ -20,29 +20,27 @@
-            KeyValueService kvs,
+    public CandidateRowsForSweepingIterator(
+            ValuesLoader valuesLoader,
             CqlExecutor cqlExecutor, TableReference table,
```
nit: this formatting is weird - either one parameter per line, or as many parameters per line as baseline allows.

Intellij does this kind of thing sometimes when you make a refactor :(

gsheasby commented on atlasdb-commons/src/main/java/com/palantir/common/base/ClosableIterator.java:
```diff
@@ -32,13 +35,29 @@
 
     default <U> ClosableIterator<U> map(Function<T, U> mapper) {
-        return ClosableIterators.wrap(stream().map(mapper).iterator(), this);
+        return ClosableIterators.wrap(Iterators.transform(this, (com.google.common.base.Function)mapper));
```
nit: baseline requires a space here: `...Function) mapper);`

We may not enforce baseline compliance for `atlasdb-commons`, but we might as well avoid introducing new issues.

gbonik commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CandidateCellForSweeping.java:
```diff
@@ -24,13 +26,13 @@
      * for the purpose of reporting the number of examined cells.
      */
-    long[] sortedTimestamps();
+    Collection<Long> sortedTimestamps();
```
Why? This means more boxing.

gbonik commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CandidateCellForSweeping.java:
```diff
@@ -41,9 +43,11 @@
-    /**
-     * The number of the (cell, timestamp) pairs examined so far.
-     */
-    long numCellsTsPairsExamined();
```
This made sense in the initial impl with in-database post-filtering, but yeah it didn't make sense anymore after the simplification

gbonik commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * We always finish the last whole row when fetching timestamps. Sweep actually only requires that we return a
```
What if the last row is wide and does not fit in memory? Seems like this implementation is not OOM-proof. You need to page through the remainder of the last row as well.

nziebart replied:
We did this previously with AllCellsPerRowPager, so I left the bahvior as is

gbonik replied:
Fair point, however the implementation that you reverted avoided that.

gbonik commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetEmptyLatestValues.java:
```diff
@@ -0,0 +1,72 @@
+                .map(CellWithTimestamps::cell)
+                .collect(Collectors.toSet());
+
+        Map<Cell, Value> valuesByCell = valuesLoader.getValues(table, cells, maxTimestampExclusive);
```
...So we still load all values if the strategy is THOROUGH? What is exactly the point of this change then?

Also, have you seen the implementation of loadWithTs? (Which is the code behind the ValuesLoader) It makes a separate request for each unique column name. For wide rows, this code will perform very poorly.

nziebart replied:
The difference is that we only load values for the latest cell (as opposed to every cell).

ValuesLoader was meant to be a conservative impl pending benchmarks. The main goal was to not OOM or overfetch. But agree that we should revisit this.

gbonik replied:
But in the case of an (almost) clean table, every cell will basically be the latest cell. Which means that in the worst case, the OOM concern is still valid. (And this worst case is pretty common - most tables should be almost clean if sweep is doing its job well).

Also, internally Cassandra will still fetch all values from disk, so we are not saving much by trying to load the latest values separately.

(By the way, you can run benchmarks manually before you merge, or even push a change - it's pretty easy, ping me on Slack if you need help)

gbonik commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCandidateRowsForSweeping.java:
```diff
@@ -0,0 +1,114 @@
+            return;
+        }
+
+        cellsWithEmptyValues = new GetEmptyLatestValues(cellTimestamps, valuesLoader, table,
```
This is technically incorrect. Since you load "cellsWithEmptyValues" in a separate request, the latest timestamps at that point might not match the latest timestamps you got during fetchCellTimestamps().

gbonik replied:
Also this is an extra round trip to Cassandra. If you are loading values anyways, why not do it during the first call.

nziebart replied:
Note that this was also the case before (in GetCandidateCellsForSweepingShim). My assumption was that it is not actually I incorrect; if the latest value is empty, it's ok to delete any previous value.

nziebart replied:
The problem isn't that you cannot load just the latest value

gbonik replied:
Yes this was the case before (and is a known problem), but why keep the bug if we can get rid of it.

> if the latest value is empty, it's ok to delete any previous value.
I don't think that's true. You need to first check if the deleted value is committed.

gbonik commented:
To summarize, I don't think this change is an improvement. I still loads values in memory in the THOROUGH case, but introduces quite a few serious problems.

In fact, the first benchmark results just came in:
fullTableScanOneWideRowThorough now takes 450 seconds over 2.6 in my impl (that's about 173x slowdown). This is quite expected from my comment about loadWithTs.
fullTableScanCleanThorough now takes 70 seconds instead of 10 seconds (7x slowdown)

Thorough strategy and wide rows on Cassandra are an actual use case at one of our government customers. And with the recent discussion on cleaning up sentinels, don't we want to effectively make every table THOROUGH?

I understand that loading values in memory in not optimal if we don't need them, but I fail to see how this was a huge problem. It surely used to be a problem before, when we loaded whole partitions in memory and OOMed. But my implementation guaranteed that we never load more than a given number of values at once. So if there is a reasonable upper bound of the value size (say, a few megabytes), and we use a batch size of 100, we'd never load more than a few hundred megs at once. If people have larger values, they should probably store them differently anyway (e.g. use the stream store feature).

On the other hand, you implementation has unbounded memory usage in the wide row scenario, and the THOROUGH case is super inefficient (as proven by the benchmarks).

gsheasby commented:
Given the problems we are seeing in multiple places, it may be best to prioritise concerns in the following order:
1 - don't OOM (or don't put undue pressure on Cassandra)
2 - performance for tables that have lots of cruft
3 - performance for clean tables

Other priorities mean that we might only have time for point 1 from the above in the next three weeks.
